### PR TITLE
`Development`: Adapt LTI services to align with centralized configuration management 

### PR DESCRIPTION
**Pull request by basak-akan**
#### General

- [x] I tested **all** changes and their related features with **all** corresponding user types on a test server.
- [x] Language: I followed the [guidelines for inclusive, diversity-sensitive, and appreciative language](https://docs.artemis.cit.tum.de/dev/guidelines/language-guidelines/).
- [x] I chose a title conforming to the [naming conventions for pull requests](https://docs.artemis.cit.tum.de/dev/development-process/#naming-conventions-for-github-pull-requests).
#### Server
- [x] **Important**: I implemented the changes with a very good performance and prevented too many (unnecessary) database calls.
- [x] I followed the [coding and design guidelines](https://docs.artemis.cit.tum.de/dev/guidelines/server/).
- [x] I added multiple integration tests (Spring) related to the features (with a high test coverage).
- [x] I added pre-authorization annotations according to the [guidelines](https://docs.artemis.cit.tum.de/dev/guidelines/server/#rest-endpoint-best-practices-for-authorization) and checked the course groups for all new REST Calls (security).
- [x] I documented the Java code using JavaDoc style.
#### Client
- [x] **Important**: I implemented the changes with a very good performance, prevented too many (unnecessary) REST calls and made sure the UI is responsive, even with large data.
- [x] I followed the [coding and design guidelines](https://docs.artemis.cit.tum.de/dev/guidelines/client/).
- [x] Following the [theming guidelines](https://docs.artemis.cit.tum.de/dev/guidelines/client-design/), I specified colors only in the theming variable files and checked that the changes look consistent in both the light and the dark theme.
- [x] I added multiple integration tests (Jest) related to the features (with a high test coverage), while following the [test guidelines](https://docs.artemis.cit.tum.de/dev/guidelines/client-tests/).
- [x] I added `authorities` to all new routes and checked the course groups for displaying navigation elements (links, buttons).
- [x] I documented the TypeScript code using JSDoc style.
- [x] I added multiple screenshots/screencasts of my UI changes.
- [x] I translated all newly inserted strings into English and German.

### Motivation and Context

As a follow-up PR for #7620, all remaining Lti 1.3 services are adapted for new approach. 

### Description

- Admins can select external platform configuration for each Online Course.
- Online course needs external platform selected to be available for external LMS (e.g. Moodle).
- Instructors can select courses and then exercises through deep linking flow.
- Exercise results and feedback are only sent to the configured platform of the relevant online course.
- Course Management --> LTI Configuration page moved to a nav bar and page re-arranged for better usability.

### Steps for Testing

Prerequisites:
- Please use TS11 for this PR.

1. Login to Moodle (https://moodle.ase.in.tum.de/) using instructor credentials (e.g. artemis_test_user_1)
2. Select "LTI Testing Course"
3. Enable edit mode (on top right corner of the page)
4. Select add an activity or resource
5. You will be directed to editing external tool page
6. Under preconfigured tools select "Artemis - https://artemis-test11.artemis.in.tum.de" and click "select content" button.
7. Verify you see login page of artemis and login using same instructor credentials.
8. Verıfy you see "LTI Single Configuration Test" course and select this course
9. Select one of the exercises and click on "link" button
10. Verify the exercise link copied to Moodle and click "save and display"
11. Verify you see linked Artemis exercise in Moodle iframe
12. Log out from Moodle and login using any of the student credentials.
13. Go to "LTI Testing Course" and click on recently added exercise
14. Participate exercise.
15.  If it is a programming exercise/quiz exercise, go to grades page on Moodle and verify you see the result.
16.  else in a incognito tab login artemis as instructor and go to course management  "LTI Single Configuration Test" course and click the exercise you added, go to Scores page and asses your participation save and submit.
17. Go back to Moodle page where you signed in as a student, go to Moodle grades page and verify you see the result and feedback.

#### Performance Review
- [x] I (as a reviewer) confirm that the client changes (in particular related to REST calls and UI responsiveness) are implemented with a very good performance 
- [x] I (as a reviewer) confirm that the server changes (in particular related to database calls) are implemented with a very good performance
#### Code Review
- [x] Code Review 1
- [x] Code Review 2
#### Manual Tests
- [x] Test 1
- [x] Test 2

#### Client

| Class/File | Line Coverage | Confirmation (assert/expect) |
|------------|--------------:|---------------------:|
| [edit-lti-configuration.component.ts](https://bamboo.ase.in.tum.de/browse/ARTEMIS-TESTS5822/latest/artifact/shared/Coverage-Report-Client-Tests/app/admin/lti-configuration/edit-lti-configuration.component.ts.html) | 96.66% | ✅ |
| [lti-configuration.component.ts](https://bamboo.ase.in.tum.de/browse/ARTEMIS-TESTS5822/latest/artifact/shared/Coverage-Report-Client-Tests/app/admin/lti-configuration/lti-configuration.component.ts.html) | 97.05% | ✅ |
| [lti-configuration.service.ts](https://bamboo.ase.in.tum.de/browse/ARTEMIS-TESTS5822/latest/artifact/shared/Coverage-Report-Client-Tests/app/admin/lti-configuration/lti-configuration.service.ts.html) | 100% | ✅ |
| [course-lti-configuration.component.ts](https://bamboo.ase.in.tum.de/browse/ARTEMIS-TESTS5822/latest/artifact/shared/Coverage-Report-Client-Tests/app/course/manage/course-lti-configuration/course-lti-configuration.component.ts.html) | 96.77% | ✅ |
| [edit-course-lti-configuration.component.ts](https://bamboo.ase.in.tum.de/browse/ARTEMIS-TESTS5822/latest/artifact/shared/Coverage-Report-Client-Tests/app/course/manage/course-lti-configuration/edit-course-lti-configuration.component.ts.html) | 82.5% | ✅ |
| [course-management-tab-bar.component.ts](https://bamboo.ase.in.tum.de/browse/ARTEMIS-TESTS5822/latest/artifact/shared/Coverage-Report-Client-Tests/app/course/manage/course-management-tab-bar/course-management-tab-bar.component.ts.html) | 94.8% | ✅ |
| [course-management.service.ts](https://bamboo.ase.in.tum.de/browse/ARTEMIS-TESTS5822/latest/artifact/shared/Coverage-Report-Client-Tests/app/course/manage/course-management.service.ts.html) | 86.14% | ✅ |
| [online-course-configuration.model.ts](https://bamboo.ase.in.tum.de/browse/ARTEMIS-TESTS5822/latest/artifact/shared/Coverage-Report-Client-Tests/app/entities/online-course-configuration.model.ts.html) | 100% | ✅ |
| [lti-course-card.component.ts](https://bamboo.ase.in.tum.de/browse/ARTEMIS-TESTS5822/latest/artifact/shared/Coverage-Report-Client-Tests/app/lti/lti-course-card.component.ts.html) | 100% | ✅ |
| [lti13-dynamic-registration.component.ts](https://bamboo.ase.in.tum.de/browse/ARTEMIS-TESTS5822/latest/artifact/shared/Coverage-Report-Client-Tests/app/lti/lti13-dynamic-registration.component.ts.html) | 100% | ✅ |
| [lti13-exercise-launch.component.ts](https://bamboo.ase.in.tum.de/browse/ARTEMIS-TESTS5822/latest/artifact/shared/Coverage-Report-Client-Tests/app/lti/lti13-exercise-launch.component.ts.html) | 95.52% | ✅ |
| [lti13-select-course.component.ts](https://bamboo.ase.in.tum.de/browse/ARTEMIS-TESTS5822/latest/artifact/shared/Coverage-Report-Client-Tests/app/lti/lti13-select-course.component.ts.html) | 92.3% | ✅ |
| [navbar.component.ts](https://bamboo.ase.in.tum.de/browse/ARTEMIS-TESTS5822/latest/artifact/shared/Coverage-Report-Client-Tests/app/shared/layouts/navbar/navbar.component.ts.html) | 88.81% | ✅ |

#### Server

| Class/File | Line Coverage | Confirmation (assert/expect) |
|------------|--------------:|---------------------:|
| [LtiPlatformConfiguration.java](https://bamboo.ase.in.tum.de/browse/ARTEMIS-TESTS5822/latest/artifact/shared/Coverage-Report-Server-Tests/de.tum.in.www1.artemis.domain/LtiPlatformConfiguration.html) | 92% | ✅ |
| [OnlineCourseConfiguration.java](https://bamboo.ase.in.tum.de/browse/ARTEMIS-TESTS5822/latest/artifact/shared/Coverage-Report-Server-Tests/de.tum.in.www1.artemis.domain/OnlineCourseConfiguration.html) | 100% | ✅ |
| [Lti13ClientRegistration.java](https://bamboo.ase.in.tum.de/browse/ARTEMIS-TESTS5822/latest/artifact/shared/Coverage-Report-Server-Tests/de.tum.in.www1.artemis.domain.lti/Lti13ClientRegistration.html) | 84% | ✅ |
| [CourseRepository.java](https://bamboo.ase.in.tum.de/browse/ARTEMIS-TESTS5822/latest/artifact/shared/Coverage-Report-Server-Tests/de.tum.in.www1.artemis.repository/CourseRepository.html) | 78% | ✅ |
| [LtiPlatformConfigurationRepository.java](https://bamboo.ase.in.tum.de/browse/ARTEMIS-TESTS5822/latest/artifact/shared/Coverage-Report-Server-Tests/de.tum.in.www1.artemis.repository/LtiPlatformConfigurationRepository.html) | 76% | ✅ |
| OnlineCourseConfigurationRepository.java | not found (modified) | ❌ |
| [Lti13LaunchFilter.java](https://bamboo.ase.in.tum.de/browse/ARTEMIS-TESTS5822/latest/artifact/shared/Coverage-Report-Server-Tests/de.tum.in.www1.artemis.security.lti/Lti13LaunchFilter.html) | 100% | ✅ |
| [CourseService.java](https://bamboo.ase.in.tum.de/browse/ARTEMIS-TESTS5822/latest/artifact/shared/Coverage-Report-Server-Tests/de.tum.in.www1.artemis.service/CourseService.html) | 89% | ✅ |
| [OnlineCourseConfigurationService.java](https://bamboo.ase.in.tum.de/browse/ARTEMIS-TESTS5822/latest/artifact/shared/Coverage-Report-Server-Tests/de.tum.in.www1.artemis.service/OnlineCourseConfigurationService.html) | 95% | ✅ |
| [Lti13Service.java](https://bamboo.ase.in.tum.de/browse/ARTEMIS-TESTS5822/latest/artifact/shared/Coverage-Report-Server-Tests/de.tum.in.www1.artemis.service.connectors.lti/Lti13Service.html) | 91% | ✅ |
| [LtiDeepLinkingService.java](https://bamboo.ase.in.tum.de/browse/ARTEMIS-TESTS5822/latest/artifact/shared/Coverage-Report-Server-Tests/de.tum.in.www1.artemis.service.connectors.lti/LtiDeepLinkingService.html) | 96% | ✅ |
| [LtiDynamicRegistrationService.java](https://bamboo.ase.in.tum.de/browse/ARTEMIS-TESTS5822/latest/artifact/shared/Coverage-Report-Server-Tests/de.tum.in.www1.artemis.service.connectors.lti/LtiDynamicRegistrationService.html) | 100% | ✅ |
| [LtiService.java](https://bamboo.ase.in.tum.de/browse/ARTEMIS-TESTS5822/latest/artifact/shared/Coverage-Report-Server-Tests/de.tum.in.www1.artemis.service.connectors.lti/LtiService.html) | 97% | ✅ |
| [CourseResource.java](https://bamboo.ase.in.tum.de/browse/ARTEMIS-TESTS5822/latest/artifact/shared/Coverage-Report-Server-Tests/de.tum.in.www1.artemis.web.rest/CourseResource.html) | 94% | ✅ |
| [LtiResource.java](https://bamboo.ase.in.tum.de/browse/ARTEMIS-TESTS5822/latest/artifact/shared/Coverage-Report-Server-Tests/de.tum.in.www1.artemis.web.rest/LtiResource.html) | 74% | ✅ |
| [AdminLtiConfigurationResource.java](https://bamboo.ase.in.tum.de/browse/ARTEMIS-TESTS5822/latest/artifact/shared/Coverage-Report-Server-Tests/de.tum.in.www1.artemis.web.rest.admin/AdminLtiConfigurationResource.html) | 87% | ✅ |
| [OnlineCourseDTO.java](https://bamboo.ase.in.tum.de/browse/ARTEMIS-TESTS5822/latest/artifact/shared/Coverage-Report-Server-Tests/de.tum.in.www1.artemis.web.rest.dto/OnlineCourseDTO.html) | 100% | ✅ |

### Screenshots

New LTI Configuration nav bar:
image

When platform is not selected it shows a warning symbol with a tool tip:
image

Admin can change Lti 1.3 configured platform:
image

After platform is configured and view advanced settings enable:
image

Select configured online course from Moodle iframe:
image

## Summary by CodeRabbit

- **New Features**
  - Introduced LTI 1.3 configuration section for administrators in the course management interface.
  - Added new LTI course card components for displaying course information within LTI applications.
  - Implemented new methods for retrieving online courses tailored for LTI dashboards.

- **Enhancements**
  - Enhanced the existing LTI integration by refining deep linking URL construction.
  - Improved the user experience by adding a new tab for LTI Configuration in the course management interface.
  - Optimized the fetching of online courses with specific LTI registration IDs.

- **Bug Fixes**
  - Resolved an issue where uninitialized online course configurations could cause errors.
  - Updated the logout process in LTI services to ensure a smoother user experience.

- **Refactor**
  - Streamlined the handling of LTI result submissions and deep linking logic.
  - Refactored LTI-related components for better maintainability and readability.

- **Documentation**
  - Updated comments and documentation related to LTI configuration and management.

- **Tests**
  - Expanded test coverage for LTI platform configurations and online course retrieval.
  - Introduced new test components for LTI course cards and selection.

- **Style**
  - Improved UI elements related to LTI configuration in the course management section.

- **Chores**
  - Removed outdated LTI configuration sections from certain components.
  - Cleaned up imports and code organization in LTI-related modules and components.

### CODE REVIEW

File: src/main/webapp/app/lti/lti.route.ts
Comment (milljoniaer) [inline src/main/webapp/app/lti/lti.route.ts:34]: I am not 100% sure if this resolves your problem, but you are missing authorities and the `canDeactivate` prop here. (Does the header work correctly for the `Lti13DeepLinkingComponent` specified below?)

You can probably copy some of the configuration of the Course Managament route:
https://github.com/ls1intum/Artemis/blob/3f0878371d18ccba1d991febe15254b142f280fc/src/main/webapp/app/course/manage/course-management.route.ts#L27-L35

Comment (basak-akan) [reply src/main/webapp/app/lti/lti.route.ts:34]: Thank you so much for your help! Adding `canActivate: [UserRouteAccessService]` fixed my issue.
Actually, the header is working for `Lti13DeepLinkingComponent`.

**Review by lennart-keller — APPROVED** [review]
Tested on ts11 in testing session. Worked well

**Review by rriyaldhi — APPROVED** [review]
Tested during the testing session and works as expected and code also looks good.

**Review by xHadie — DISMISSED** [review]
_Manually tested on TS11_Legacy_
Works as expected, thanks for the clear testing instructions!

**Review by MaximilianJG — DISMISSED** [review]
Tested in a testing session. Works as expected!

**Review by coolchock — APPROVED** [review]
tested in testing session, works as expected

**Review by egekurt123 — DISMISSED** [review]
Tested in a testing session, works as expected

**Review by Predixx — DISMISSED** [review]
tested during testing session. worked for me!

**Review by nata-81 — DISMISSED** [review]
tested on ts11. Works as expected.

**Review by konrad2002 — DISMISSED** [review]
manually tested on legacy ts11, works as expected

File: src/main/webapp/app/lti/lti13-select-course.component.html
Comment (basak-akan) [reply src/main/webapp/app/lti/lti13-select-course.component.html:5]: We adapted new control flow syntax in Angular 17 which also includes `@for`

File: src/main/webapp/app/lti/lti-course-card.component.html
Comment (basak-akan) [reply src/main/webapp/app/lti/lti-course-card.component.html:25]: We adapted new control flow syntax in Angular 17 which also includes `@if`

File: src/main/webapp/app/admin/lti-configuration/lti-configuration.service.ts
Comment (basak-akan) [reply src/main/webapp/app/admin/lti-configuration/lti-configuration.service.ts:20]: The used endpoint is right.

File: src/main/java/de/tum/in/www1/artemis/web/rest/admin/AdminLtiConfigurationResource.java
Comment (basak-akan) [reply src/main/java/de/tum/in/www1/artemis/web/rest/admin/AdminLtiConfigurationResource.java:1]: `getAllConfiguredLtiPlatforms` is removed from admin endpoints

**Review by rriyaldhi — APPROVED** [review]
Reapprove code

**Review by konrad2002 — DISMISSED** [review]
tested again, still works

**Review by coolchock — APPROVED** [review]
re-approve after merge conflicts were resolved

**Review by egekurt123 — DISMISSED** [review]
Reapprove after changes

**Review by lennart-keller — APPROVED** [review]
Reapprove after resolving merge conflicts

**Review by MaximilianJG — APPROVED** [review]

**Review by Mtze — DISMISSED** [review]
LGTM

**Pull request merged**
